### ISSUE TITLE
Question about Generation of Guidance Map

### ISSUE THREAD

**Issue by MeiliMa** [cocoon2wong/Vertical#2]
I am curious about the process of generating the guidance map.
I assume the guidance map will be used during model inference (rather than just training). But, I saw that the map is generated using data beyond just the agent under prediction. It is actually generated using the observed (8-frame) trajectories from all agents. (Please correctize me if I am wrong).
This is a kind of data leakage, as in reality, when do predicting, we should have no idea about how the trajectories in a scenario are distributed. Namely, you should only know what happens within exactly the 8 frames of observation.

For example, you can consider that given a 21-frame trajectory, you extract two 20-frame sub-trajectories from it and plot the first 8 frames of each sub-trajectory on the map, and then, you actually include 9 frames (1 frame more than the observation horizon) from the 1st sub-trajectory on the map. Things will become worse, if, for example, the original trajectory is 32-frame long or even longer, it will plot all the ground truth trajectory on the map.

**Comment by cocoon2wong** [comment]
The guidance map is originally designed to update dynamically when model implementing. It is used to evaluate the movable area in the scene. Details of which can be found in our previous work https://arxiv.org/abs/2010.03897. Here is a video demo for that process:
https://user-images.githubusercontent.com/107823343/180667398-ef9a9ec5-3675-460b-a653-783888cb6071.mov

Further, our another work https://www.sciencedirect.com/science/article/pii/S0031320322000334 employ a Transfer CNN to generate guidance maps from the scene RGB images on unseen scenes. Then, the obtained image will be saved as the `trajMap_load.png`, and then feed to the model like:
https://github.com/cocoon2wong/Vertical/blob/30dfcb923f7a14426a1def6bfeb7018d04824f4b/codes/dataset/__manager.py#L342-L345
It means that the guidance maps could be build either with agents' observed trajectories (if possible) and the scene images.

But note that the guidance map used in this work is actually a simplified version that builds from all observed trajectories in current scene, for the main contribution of this work lies not on these components, but the trajectory spectrums. It could be also regarded as the output of the above transferCNN with a better training.

> For example, you can consider that given a 21-frame trajectory, you extract two 20-frame sub-trajectories from it and plot the first 8 frames of each sub-trajectory on the map, and then, you actually include 9 frames (1 frame more than the observation horizon) from the 1st sub-trajectory on the map. Things will become worse, if, for example, the original trajectory is 32-frame long or even longer, it will plot all the ground truth trajectory on the map.

But I am not very clear about your example, could you please give me more descriptions about your confusion?

**Comment by MeiliMa** [comment]
Suppose that an agent has 21 frames of recorded positions in the dataset, and an interesting trajectory is supposed to have 20 frames (8-frame observation, 12-frame prediction). Then, we can extract two 20-frame sub-trajectories for that agent, i.e. one trajectory from the 1st frame to 20th frame and the other one from 2nd frame to the 21th frame. So, we can see that the 9th frame is recorded by the observation of the 2nd sub-trajectory, but is also a part of the ground-truth future trajectory for the 1st sub-trajectory to prediction. Or more general, in the guidance map, there are some overlappings between the observation and prediction data. Is it a kind of data leakage? 

Regardless of that, I do not think it is proper to introduce the observations of all agents in a testing scenario as a part of the model's input, since it is actually to introduce a biased prior, which existing approaches typically assume unknown. I did not find that you mentioned this point in your paper. 

I would strongly hope you can re-test your model without the prior and see if the performance changes.

**Comment by cocoon2wong** [comment]
> Or more general, in the guidance map, there are some overlappings between the observation and prediction data. Is it a kind of data leakage?

I know your concern. In simple terms, the guidance map acts more like a simple semantic segmentation network for images. In our previous work mentioned above, we use the guidance map C obtained by recording all trajectories in this way as the label of the output of the transferCNN, and the input of this CNN is the scene image so that we can establish the relationship between the guidance map and the scene visual image. For example, we make C on hotel, univ, and zara datasets using trajectories so that the model can predict the hat C on eth. In this work, this process is simplified by building C directly on the eth dataset in the above example, which is the interaction representations mentioned in our paper.

To further reduce your doubts, this guidance map can actually also use the scene segmentation map (already provided in the dataset), but the social maps part will still be built dynamically, and we experimented in that case with basically no significant difference in the results.

If you have reproduced some of our codes, you can find the built guidance map at ./dataset_npz/xxx/maps/xxx.png

**Comment by MeiliMa** [comment]
It makes scenes that you made the so-called `C` using only training data rather than exploiting the testing data. 
May you re-test your model without directly building `C` from the testing data? 

If you just need semantic maps, I think YNet provided those for ETH/UCY datasets, which can be found at: https://github.com/HarshayuGirase/Human-Path-Prediction/tree/master/ynet#pretrained-models-data-and-config-files

**Comment by cocoon2wong** [comment]
I ran a simple experiment. Due to video scaling and other issues (for example the ETH-UCY dataset is annotated with real-word meters), I performed a manual rough segmentation on the zara1 dataset and used this `trajMap_seg.png` to directly replace the original `trajMap.png` (at `./dataset_npz/zara1/agent1_maps`) obtained after the trajectory, and the experimental results before and after the replacement are:

with `trajMap.png`:
```none
>>> test results:
    - dataset is zara1.
    - ADE(m) is 0.20500906348228455.
    - FDE(m) is 0.3047057890892029.
```

with `trajMap_seg.png`:
```none
>>> test results:
    - dataset is zara1.
    - ADE(m) is 0.2002884203195572.
    - FDE(m) is 0.3053431930541992.
```

Test commands I used is:
```bash
python main.py --model V --loada ./weights/vertical/a_zara1 --loadb ./weights/vertical/b_universal
```

`trajMap_seg.png`:
![trajMap_seg1](https://user-images.githubusercontent.com/107823343/180669257-1cfd569a-7174-443f-bc0a-e8f73bcd9f97.png)

`trajMap.png`:
![trajMap](https://user-images.githubusercontent.com/107823343/180669279-8efdd577-885e-4612-8808-c9d4eba4cfc5.png)

The manually segmentation process：
![trajMap_seg](https://user-images.githubusercontent.com/107823343/180669287-6d034bf7-aa9f-48b4-a7e2-92152e925711.png)

**Comment by cocoon2wong** [comment]
> It makes scenes that you made the so-called `C` using only training data rather than exploiting the testing data. May you re-test your model without directly building `C` from the testing data?
> 
> If you just need semantic maps, I think YNet provided those for ETH/UCY datasets, which can be found at: https://github.com/HarshayuGirase/Human-Path-Prediction/tree/master/ynet#pretrained-models-data-and-config-files

Most importantly, the guidance map + social map will be only used in the second stage subnetwork.
The weights we pre-trained (`./weights/vertical/b_universal`) is actually trained on the SDD dataset, and it also works on the ETH-UCY:
https://github.com/cocoon2wong/Vertical/blob/30dfcb923f7a14426a1def6bfeb7018d04824f4b/weights/vertical/b_universal/args.json#L22
That subnetwork learns how to interpolate trajectory spectrums according to agents' interaction status, rather than how the scene objects distributes.

**Comment by MeiliMa** [comment]
I am confused. I run your code with the pre-trained models, and found that some of your reported numbers in the main table (Table 2) of your paper were without using `b_universal` but some, e.g. ETH, used it.  The results I obtained are inconsistent with those in your ablation study as well. I listed the results on UCY datasets, as the ETH datasets you used are different to many other baselines, though your UCY data are also not the exact same with other baselines. The following are the results I got:

| | `b_universal` | `l` |
|---|---|---|
|UNIV|  0.21/0.37 | 0.21/0.35 |
|ZARA1| 0.21/0.31 | 0.20/0.31|
|ZARA2 | 0.15/0.25 | 0.15/0.25 |

It seems that using `b_universal` (Spectrum Interpolation Sub-network, I assume) cannot help at all and even will lead to a performance discrepancy. 

Besides, those numbers are much smaller than those in the log files provided with the pre-trained models. I assume, if I set `--loadb l`, it uses only the 1st-stage trained model, whose performance should be consistent with the provided log files.

**Comment by cocoon2wong** [comment]
The '--loadb l' will use the linear interpolation instead of the original stage-2 subnetwork, and it is the variation c in ablation studies.
However i got similar results as our paper reported on my M1 mac mini. Do you have multiple runs to see the influence of random factors (the random sampling in the first-stage subnetwork)?
I will give a feedback to you later for my working computer is in my office.

**Comment by MeiliMa** [comment]
I tried running multiple times. The error occurred after the 3rd digital place and would not influence the results with 2 digital places.

May I ask what is the difference between using `--loadb l` and the log in the provided tensorboard files? Were the log files recorded during training of the model that `--loadb l` will give?

**Comment by cocoon2wong** [comment]
> I tried running multiple times. The error occurred after the 3rd digital place and would not influence the results with 2 digital places.
> 
> 
> 
> May I ask what is the difference between using `--loadb l` and the log in the provided tensorboard files? Were the log files recorded during training of the model that `--loadb l` will give?

The tensorboard files are the logs for the stage-1 subnetwork.
Two subnetworks are trained separately, please refer to the README.

**Comment by cocoon2wong** [comment]
> I tried running multiple times. The error occurred after the 3rd digital place and would not influence the results with 2 digital places.

The most likely reason is probably because I uploaded the wrong weights for the second stage subnetwork, please wait for me to check later.

**Comment by cocoon2wong** [comment]
> > I tried running multiple times. The error occurred after the 3rd digital place and would not influence the results with 2 digital places.
> 
> The most likely reason is probably because I uploaded the wrong weights for the second stage subnetwork, please wait for me to check later.

@MeiliMa I think I have found the cause of the problem. The preprocessing parameters of the stage-2 model in this repository is currently set incorrectly, which is different from my local codes. Specifically, the reference point of the `moving` should be the last point of the observed trajectory instead of the first point. This problem arises here:
https://github.com/cocoon2wong/Vertical/blob/30dfcb923f7a14426a1def6bfeb7018d04824f4b/silverballers/handlers/__baseHandler.py#L19-L33
It should be the 
```python
        # Preprocess
        self.set_preprocess('move')
        self.set_preprocess_parameters(move=-1)
```
The results in the paper were all obtained using `move=-1`. I will sort out our weights and upload them later to replace the incorrect `b_universal`. (the `b_universal` was trained under `move=0`)

Please note that both extra data preprocess methods and the silverballers structures are not included in our conference paper, and the `V^2-Net` is a special case of this structure. Current `V^2-Net` codes are actually re-written with that silver ballers structure. We are preparing the corresponding  journal version of the paper with expanded details. If you are interested, feel free to check out our [development repository branch](https://github.com/cocoon2wong/Project-Jocker/tree/bb).

**Comment by cocoon2wong** [comment]
@MeiliMa This issue will be automatically closed once the [pull request#3](https://github.com/cocoon2wong/Vertical/pull/3) is merged. If you have any other questions, please start a new issue😊
